### DISCUSSION TITLE
171 Modularization, Managing SPM Targets & Dependencies

### DISCUSSION THREAD

**Discussion by JongHyunLee84** [pointfreeco/episode-code-samples#153] [Q&A] [answered]
I'm leaving a question in the Discussion because there aren't many references for Local SPM.

In the Package.swift file below, both 'AppFeature' and 'InventoryFeature' have 'Model' target as a dependency.

In this case, does AppFeature, which has InventoryFeature as a dependency, copy Model twice redundantly? I'm confused due to the mixed concepts of static and dynamic libraries.

Additionally, what's the reason for creating .library for each target in products? Is it to create a buildable scheme in Xcode by adding it to products?

```
let package = Package(
  name: "Inventory",
  platforms: [.iOS(.v15)],
  products: [
    .library(name: "AppFeature", targets: ["AppFeature"]),
    .library(name: "InventoryFeature", targets: ["InventoryFeature"]),
    .library(name: "ItemFeature", targets: ["ItemFeature"]),
    .library(name: "ItemRowFeature", targets: ["ItemRowFeature"]),
    .library(name: "Models", targets: ["Models"]),
    .library(name: "ParsingHelpers", targets: ["ParsingHelpers"]),
    .library(name: "SwiftUIHelpers", targets: ["SwiftUIHelpers"])
  ],
  dependencies: [
    .package(url: "https://github.com/pointfreeco/swift-case-paths", from: "0.7.0"),
    .package(url: "https://github.com/pointfreeco/swift-identified-collections", from: "0.3.2"),
    .package(url: "https://github.com/pointfreeco/swift-parsing", from: "0.3.1"),
  ],
  targets: [
    .target(
      name: "AppFeature",
      dependencies: [
        "InventoryFeature",
        "Models",
        "ParsingHelpers",
        .product(name: "Parsing", package: "swift-parsing"),
      ]
    ),
    .target(
      name: "InventoryFeature",
      dependencies: [
        "ItemRowFeature",
        "Models",
        "SwiftUIHelpers",
        .product(name: "CasePaths", package: "swift-case-paths"),
        .product(name: "IdentifiedCollections", package: "swift-identified-collections"),
      ]
    ),
    ....
```

**Comment by mbrandonw** [comment]
> In this case, does AppFeature, which has InventoryFeature as a dependency, copy Model twice redundantly? I'm confused due to the mixed concepts of static and dynamic libraries.

Hi @JongHyunLee84, we prefer to be explicit with what modules a feature depends on rather than using a dependency's transitive dependencies (like InventoryFeature's Models). You can drop the Models from the AppFeature if you want, but I believe someday in the future SPM may become more strict in not allowing access to transitive dependencies like that.

> Additionally, what's the reason for creating .library for each target in products? Is it to create a buildable scheme in Xcode by adding it to products?

Yes exactly, it is so that we can build each of those libraries in isolation in Xcode.

**Reply by JongHyunLee84** [reply]
@mbrandonw 
Thank you for your answer.
Then, even though both AppFeature and InventoryFeature explicitly depend on Model, does this not cause any issues such as Model being used inefficiently twice?

**Reply by mihaicris-adoreme** [reply]
I think SPM makes some magic and enforces that the Model target should be a dynamic framework and not static.

**Reply by mbrandonw** [reply, marked as answer]
Yes, I believe what @mihaicris-adoreme said is correct, and I believe that has been part of the difficult work that Tuist has dealt with in trying to figure out Xcode+SPM's magic.
